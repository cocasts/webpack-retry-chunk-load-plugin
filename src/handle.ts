/* eslint-disable */
// @ts-nocheck
import * as prettier from 'prettier';

const pluginName = 'HandleChunkLoadFailurePlugin';

/**
 * This Webpack plugin calls `window._chunkHandler.handleVersionChange()` when loading a chunk fails
 */
export class HandleChunkLoadFailurePlugin {
  constructor(options = {}) {
    this.options = Object.assign({}, options);
  }

  apply(compiler) {
    compiler.hooks.compilation.tap(pluginName, compilation => {
      const { mainTemplate } = compilation;
      if (mainTemplate.hooks.jsonpScript) {
        // Adapted from https://github.com/webpack/webpack/blob/11e94dd2d0a8d8baae75e715ff8a69f27a9e3014/lib/web/JsonpMainTemplatePlugin.js#L145-L210
        // https://github.com/webpack/webpack/issues/9611#issuecomment-523850870
        mainTemplate.hooks.jsonpScript.tap(pluginName, (source, chunk) => {
          const { crossOriginLoading, chunkLoadTimeout, jsonpScriptType } =
            mainTemplate.outputOptions;

          const crossOriginScript = `
          if (script.src.indexOf(window.location.origin + '/') !== 0) {
            script.crossOrigin = ${JSON.stringify(crossOriginLoading)};
          }
        `;

          const scriptWithRetry = `
          // create error before stack unwound to get useful stacktrace later
          var error = new Error();
          function loadScript(src) {
            var script = document.createElement('script');
            var onScriptComplete;
            ${
              jsonpScriptType
                ? `script.type = ${JSON.stringify(jsonpScriptType)};`
                : ''
            }
            script.charset = 'utf-8';
            script.timeout = ${chunkLoadTimeout / 1000};
            if (${mainTemplate.requireFn}.nc) {
              script.setAttribute("nonce", ${mainTemplate.requireFn}.nc);
            }
            script.src = src;
            ${crossOriginLoading ? crossOriginScript : ''}
            onScriptComplete = function (event) {
              var chunk = installedChunks[chunkId];
              if (chunk) window._chunkHandler.handleVersionChange()
            };
            script.onerror = onScriptComplete;
            return script;
          }
          var script = loadScript(jsonpScriptSrc(chunkId));
        `;

          const currentChunkName = chunk.name;
          const addRetryCode =
            !this.options.chunks ||
            this.options.chunks.includes(currentChunkName);
          const script = addRetryCode ? scriptWithRetry : source;

          return prettier.format(script, {
            singleQuote: true,
            parser: 'babel',
          });
        });
      }
    });
  }
}

// module.exports[pluginName] = HandleChunkLoadFailurePlugin;
