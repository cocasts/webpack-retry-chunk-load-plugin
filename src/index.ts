import * as prettier from 'prettier';
import { Compiler, RuntimeGlobals } from 'webpack';

const pluginName = 'RetryChunkLoadPlugin';

export interface RetryChunkLoadPluginOptions {
  /**
   * optional stringified function to get the cache busting query string appended to the script src
   * if not set will default to appending the string `?cache-bust=true`
   */
  cacheBust?: string;
  /**
   * optional list of chunks to which retry script should be injected
   * if not set will add retry script to all chunks that have webpack script loading
   */
  chunks?: string[];
  /**
   * optional code to be executed in the browser context if after all retries chunk is not loaded.
   * if not set - nothing will happen and error will be returned to the chunk loader.
   */
  lastResortScript?: string;
  /**
   * optional value to set the maximum number of retries to load the chunk. Default is 1
   */
  maxRetries?: number;
  /**
   * optional number value to set the amount of time in milliseconds before trying to load the chunk again. Default is 0
   * if string, value must be code to generate a delay value. Receives retryCount as argument
   * e.g. `function(retryAttempt) { return retryAttempt * 1000 }`
   */
  retryDelay?: number | string;
  /**
   * optional string value to load chunk using fallback path
   * if set will try one more time after max retries failed
   */
  fallbackPath?: string;
}

export class RetryChunkLoadPlugin {
  options: RetryChunkLoadPluginOptions;

  constructor(options: RetryChunkLoadPluginOptions = {}) {
    this.options = Object.assign({}, options);
  }

  apply(compiler: Compiler) {
    compiler.hooks.thisCompilation.tap(pluginName, compilation => {
      const { mainTemplate, runtimeTemplate } = compilation;
      const maxRetryValueFromOptions = Number(this.options.maxRetries);
      const maxRetries =
        Number.isInteger(maxRetryValueFromOptions) &&
        maxRetryValueFromOptions > 0
          ? maxRetryValueFromOptions
          : 1;
      const fallbackPath = this.options.fallbackPath;
      const shouldTryFallback = !!fallbackPath;
      const getCacheBustString = () =>
        this.options.cacheBust
          ? `
                  (${this.options.cacheBust})();
                `
          : '"cache-bust=true"';
      mainTemplate.hooks.localVars.tap(
        { name: pluginName, stage: 1 },
        (source, chunk) => {
          const currentChunkName = chunk.name;
          const addRetryCode =
            !this.options.chunks ||
            this.options.chunks.includes(currentChunkName);

          const getRetryDelay =
            typeof this.options.retryDelay === 'string'
              ? this.options.retryDelay
              : `function() { return ${this.options.retryDelay || 0} }`;

          if (!addRetryCode) return source;
          const script = runtimeTemplate.iife(
            '',
            `
          if(typeof ${RuntimeGlobals.require} !== "undefined") {
            var oldGetScript = ${RuntimeGlobals.getChunkScriptFilename};
            var oldLoadScript = ${RuntimeGlobals.loadScript};
            var oldEnsureScript = ${RuntimeGlobals.ensureChunk};
            var queryMap = {};
            var countMap = {};
            var isTriedFallbackMap = {};
            var getRetryDelay = ${getRetryDelay};
            var fallbackPath = '${fallbackPath}';

            ${RuntimeGlobals.loadScript} = function(url, done, key, chunkId) {
              var retries = countMap.hasOwnProperty(chunkId) ? countMap[chunkId] : ${maxRetries};
              var finalUrl = url;
              if (retries < 0 && ${shouldTryFallback}) {
                finalUrl = url.replace(${
                  RuntimeGlobals.publicPath
                }, fallbackPath);
                isTriedFallbackMap[chunkId] = true;
              }
              var result = oldLoadScript(finalUrl, done, key, chunkId);
              return result;
            };
            ${RuntimeGlobals.getChunkScriptFilename} = function(chunkId){
              var result = oldGetScript(chunkId);
              return result + (queryMap.hasOwnProperty(chunkId) ? '?' + queryMap[chunkId]  : '');
            };
            ${RuntimeGlobals.ensureChunk} = function(chunkId){
              var result = oldEnsureScript(chunkId);
              return result.catch(function(error){
                var retries = countMap.hasOwnProperty(chunkId) ? countMap[chunkId] : ${maxRetries};
                if (retries < 1 && (${shouldTryFallback} === !!isTriedFallbackMap[chunkId])) {
                  var realSrc = oldGetScript(chunkId);
                  error.message = 'Loading chunk ' + chunkId + ' failed after ${maxRetries} retries.\\n(' + realSrc + ')';
                  error.request = realSrc;${
                    this.options.lastResortScript
                      ? this.options.lastResortScript
                      : ''
                  }
                  
                  throw error;
                }
                return new Promise(function (resolve) {
                  var retryAttempt = ${maxRetries} - retries + 1;
                  setTimeout(function () {
                    var retryAttemptString = '&retry-attempt=' + retryAttempt;
                    var cacheBust = ${getCacheBustString()} + retryAttemptString;
                    queryMap[chunkId] = cacheBust;
                    countMap[chunkId] = retries - 1;
                    resolve(${RuntimeGlobals.ensureChunk}(chunkId));
                  }, getRetryDelay(retryAttempt))
                })
              });
            };
          }`
          );
          return (
            source +
            prettier.format(script, {
              trailingComma: 'es5',
              singleQuote: true,
              parser: 'babel',
            })
          );
        }
      );
    });
  }
}
