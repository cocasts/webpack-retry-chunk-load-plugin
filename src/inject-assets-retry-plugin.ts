import { readFileSync } from 'fs';
import HtmlWebpackPlugin = require('html-webpack-plugin');
import { Compiler } from 'webpack';

const pluginName = 'InjectAssetsRetryPlugin';

export interface AssetsRetryOptions {
  maxRetryCount?: number;
  onRetry?: RetryFunction;
  onSuccess?: SuccessFunction;
  onFail?: FailFunction;
  domain?: Domain;
}
export type RetryFunction = (
  currentUrl: string,
  originalUrl: string,
  retryCollector: null | RetryStatistics
) => string | null;
export interface RetryStatistics {
  retryTimes: number;
  succeeded: string[];
  failed: string[];
}
export type SuccessFunction = (currentUrl: string) => void;
export type FailFunction = (currentUrl: string) => void;
export type Domain = string[] | { [x: string]: string };

export class InjectAssetsRetryPlugin {
  options: AssetsRetryOptions;

  constructor(options: AssetsRetryOptions) {
    this.options = Object.assign({}, options);
  }

  apply(compiler: Compiler) {
    compiler.hooks.compilation.tap(pluginName, compilation => {
      const hooks = HtmlWebpackPlugin.getHooks(compilation);
      hooks.alterAssetTagGroups.tap(
        {
          name: pluginName,
          stage: 1,
        },
        assets => {
          assets.headTags.unshift({
            attributes: {},
            tagName: 'script',
            innerHTML: `
            window.assetsRetry({
              domain: ${JSON.stringify(this.options.domain)},
              maxRetryCount: ${this.options.maxRetryCount},
              onRetry: ${this.options.onRetry},
              onSuccess: ${this.options.onSuccess},
              onFail: ${this.options.onFail},
            });
          `,
            voidTag: false,
            meta: { plugin: pluginName },
          });
          assets.headTags.unshift({
            attributes: {},
            tagName: 'script',
            innerHTML: readFileSync(require.resolve('assets-retry')).toString(),
            voidTag: false,
            meta: { plugin: pluginName },
          });
          return assets;
        }
      );
    });
  }
}
