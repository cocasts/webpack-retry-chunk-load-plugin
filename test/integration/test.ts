import * as path from 'path';
import { RetryChunkLoadPluginOptions } from '../../src';
import webpack from './utils/webpack';

const mainOutputFile = path.join(__dirname, 'fixtures', 'dist', 'main.js');

const cases: (RetryChunkLoadPluginOptions | undefined)[] = [
  undefined,
  { chunks: [] },
  { chunks: ['main'] },
  { chunks: ['main'], retryDelay: 3000 },
  {
    chunks: ['main'],
    retryDelay: 'function(retryAttempt) { return retryAttempt * 1000 }',
  },
  {
    chunks: ['main'],
    lastResortScript: "window.location.href='/500.html'",
  },
];

test.each(cases)('given config %j, match snapshot', async config => {
  const { result, fs } = webpack(config);
  await result;
  const mainContents = fs.readFileSync(mainOutputFile).toString();
  expect(mainContents).toMatchSnapshot();
});
