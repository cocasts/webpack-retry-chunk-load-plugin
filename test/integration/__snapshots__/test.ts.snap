// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given config {"chunks":["main"],"lastResortScript":"window.location.href='/500.html'"}, match snapshot 1`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	\\"use strict\\";
/******/ 	var __webpack_modules__ = ({

/***/ \\"./test/integration/fixtures/async.ts\\":
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/async.ts ***!
  \\\\********************************************/
/***/ ((__unused_webpack_module, exports) => {


Object.defineProperty(exports, \\"__esModule\\", ({ value: true }));
exports.foo = void 0;
exports.foo = 'foo';
console.log('foo loaded!');


/***/ })

/******/ 	});
/************************************************************************/
/******/ 	// The module cache
/******/ 	var __webpack_module_cache__ = {};
/******/ 	
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/ 		// Check if module is in cache
/******/ 		var cachedModule = __webpack_module_cache__[moduleId];
/******/ 		if (cachedModule !== undefined) {
/******/ 			return cachedModule.exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = __webpack_module_cache__[moduleId] = {
/******/ 			// no module.id needed
/******/ 			// no module.loaded needed
/******/ 			exports: {}
/******/ 		};
/******/ 	
/******/ 		// Execute the module function
/******/ 		__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/ 	
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/compat */
/******/ 	(() => {
/******/ 	  if (typeof __webpack_require__ !== 'undefined') {
/******/ 	    var oldGetScript = __webpack_require__.u;
/******/ 	    var oldLoadScript = __webpack_require__.e;
/******/ 	    var queryMap = {};
/******/ 	    var countMap = {};
/******/ 	    var getRetryDelay = function () {
/******/ 	      return 0;
/******/ 	    };
/******/ 	    __webpack_require__.u = function (chunkId) {
/******/ 	      var result = oldGetScript(chunkId);
/******/ 	      return (
/******/ 	        result +
/******/ 	        (queryMap.hasOwnProperty(chunkId) ? '?' + queryMap[chunkId] : '')
/******/ 	      );
/******/ 	    };
/******/ 	    __webpack_require__.e = function (chunkId) {
/******/ 	      var result = oldLoadScript(chunkId);
/******/ 	      return result.catch(function (error) {
/******/ 	        var retries = countMap.hasOwnProperty(chunkId) ? countMap[chunkId] : 1;
/******/ 	        if (retries < 1) {
/******/ 	          var realSrc = oldGetScript(chunkId);
/******/ 	          error.message =
/******/ 	            'Loading chunk ' +
/******/ 	            chunkId +
/******/ 	            ' failed after 1 retries.\\\\n(' +
/******/ 	            realSrc +
/******/ 	            ')';
/******/ 	          error.request = realSrc;
/******/ 	          window.location.href = '/500.html';
/******/ 	          throw error;
/******/ 	        }
/******/ 	        return new Promise(function (resolve) {
/******/ 	          var retryAttempt = 1 - retries + 1;
/******/ 	          setTimeout(function () {
/******/ 	            var retryAttemptString = '&retry-attempt=' + retryAttempt;
/******/ 	            var cacheBust = 'cache-bust=true' + retryAttemptString;
/******/ 	            queryMap[chunkId] = cacheBust;
/******/ 	            countMap[chunkId] = retries - 1;
/******/ 	            resolve(__webpack_require__.e(chunkId));
/******/ 	          }, getRetryDelay(retryAttempt));
/******/ 	        });
/******/ 	      });
/******/ 	    };
/******/ 	  }
/******/ 	})();
/******/ 	
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/index.ts ***!
  \\\\********************************************/

async function run() {
    console.log(await Promise.resolve().then(() => __webpack_require__(/*! ./async */ \\"./test/integration/fixtures/async.ts\\")));
}
run();

})();

/******/ })()
;"
`;

exports[`given config {"chunks":["main"],"retryDelay":"function(retryAttempt) { return retryAttempt * 1000 }"}, match snapshot 1`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	\\"use strict\\";
/******/ 	var __webpack_modules__ = ({

/***/ \\"./test/integration/fixtures/async.ts\\":
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/async.ts ***!
  \\\\********************************************/
/***/ ((__unused_webpack_module, exports) => {


Object.defineProperty(exports, \\"__esModule\\", ({ value: true }));
exports.foo = void 0;
exports.foo = 'foo';
console.log('foo loaded!');


/***/ })

/******/ 	});
/************************************************************************/
/******/ 	// The module cache
/******/ 	var __webpack_module_cache__ = {};
/******/ 	
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/ 		// Check if module is in cache
/******/ 		var cachedModule = __webpack_module_cache__[moduleId];
/******/ 		if (cachedModule !== undefined) {
/******/ 			return cachedModule.exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = __webpack_module_cache__[moduleId] = {
/******/ 			// no module.id needed
/******/ 			// no module.loaded needed
/******/ 			exports: {}
/******/ 		};
/******/ 	
/******/ 		// Execute the module function
/******/ 		__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/ 	
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/compat */
/******/ 	(() => {
/******/ 	  if (typeof __webpack_require__ !== 'undefined') {
/******/ 	    var oldGetScript = __webpack_require__.u;
/******/ 	    var oldLoadScript = __webpack_require__.e;
/******/ 	    var queryMap = {};
/******/ 	    var countMap = {};
/******/ 	    var getRetryDelay = function (retryAttempt) {
/******/ 	      return retryAttempt * 1000;
/******/ 	    };
/******/ 	    __webpack_require__.u = function (chunkId) {
/******/ 	      var result = oldGetScript(chunkId);
/******/ 	      return (
/******/ 	        result +
/******/ 	        (queryMap.hasOwnProperty(chunkId) ? '?' + queryMap[chunkId] : '')
/******/ 	      );
/******/ 	    };
/******/ 	    __webpack_require__.e = function (chunkId) {
/******/ 	      var result = oldLoadScript(chunkId);
/******/ 	      return result.catch(function (error) {
/******/ 	        var retries = countMap.hasOwnProperty(chunkId) ? countMap[chunkId] : 1;
/******/ 	        if (retries < 1) {
/******/ 	          var realSrc = oldGetScript(chunkId);
/******/ 	          error.message =
/******/ 	            'Loading chunk ' +
/******/ 	            chunkId +
/******/ 	            ' failed after 1 retries.\\\\n(' +
/******/ 	            realSrc +
/******/ 	            ')';
/******/ 	          error.request = realSrc;
/******/ 	          throw error;
/******/ 	        }
/******/ 	        return new Promise(function (resolve) {
/******/ 	          var retryAttempt = 1 - retries + 1;
/******/ 	          setTimeout(function () {
/******/ 	            var retryAttemptString = '&retry-attempt=' + retryAttempt;
/******/ 	            var cacheBust = 'cache-bust=true' + retryAttemptString;
/******/ 	            queryMap[chunkId] = cacheBust;
/******/ 	            countMap[chunkId] = retries - 1;
/******/ 	            resolve(__webpack_require__.e(chunkId));
/******/ 	          }, getRetryDelay(retryAttempt));
/******/ 	        });
/******/ 	      });
/******/ 	    };
/******/ 	  }
/******/ 	})();
/******/ 	
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/index.ts ***!
  \\\\********************************************/

async function run() {
    console.log(await Promise.resolve().then(() => __webpack_require__(/*! ./async */ \\"./test/integration/fixtures/async.ts\\")));
}
run();

})();

/******/ })()
;"
`;

exports[`given config {"chunks":["main"],"retryDelay":3000}, match snapshot 1`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	\\"use strict\\";
/******/ 	var __webpack_modules__ = ({

/***/ \\"./test/integration/fixtures/async.ts\\":
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/async.ts ***!
  \\\\********************************************/
/***/ ((__unused_webpack_module, exports) => {


Object.defineProperty(exports, \\"__esModule\\", ({ value: true }));
exports.foo = void 0;
exports.foo = 'foo';
console.log('foo loaded!');


/***/ })

/******/ 	});
/************************************************************************/
/******/ 	// The module cache
/******/ 	var __webpack_module_cache__ = {};
/******/ 	
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/ 		// Check if module is in cache
/******/ 		var cachedModule = __webpack_module_cache__[moduleId];
/******/ 		if (cachedModule !== undefined) {
/******/ 			return cachedModule.exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = __webpack_module_cache__[moduleId] = {
/******/ 			// no module.id needed
/******/ 			// no module.loaded needed
/******/ 			exports: {}
/******/ 		};
/******/ 	
/******/ 		// Execute the module function
/******/ 		__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/ 	
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/compat */
/******/ 	(() => {
/******/ 	  if (typeof __webpack_require__ !== 'undefined') {
/******/ 	    var oldGetScript = __webpack_require__.u;
/******/ 	    var oldLoadScript = __webpack_require__.e;
/******/ 	    var queryMap = {};
/******/ 	    var countMap = {};
/******/ 	    var getRetryDelay = function () {
/******/ 	      return 3000;
/******/ 	    };
/******/ 	    __webpack_require__.u = function (chunkId) {
/******/ 	      var result = oldGetScript(chunkId);
/******/ 	      return (
/******/ 	        result +
/******/ 	        (queryMap.hasOwnProperty(chunkId) ? '?' + queryMap[chunkId] : '')
/******/ 	      );
/******/ 	    };
/******/ 	    __webpack_require__.e = function (chunkId) {
/******/ 	      var result = oldLoadScript(chunkId);
/******/ 	      return result.catch(function (error) {
/******/ 	        var retries = countMap.hasOwnProperty(chunkId) ? countMap[chunkId] : 1;
/******/ 	        if (retries < 1) {
/******/ 	          var realSrc = oldGetScript(chunkId);
/******/ 	          error.message =
/******/ 	            'Loading chunk ' +
/******/ 	            chunkId +
/******/ 	            ' failed after 1 retries.\\\\n(' +
/******/ 	            realSrc +
/******/ 	            ')';
/******/ 	          error.request = realSrc;
/******/ 	          throw error;
/******/ 	        }
/******/ 	        return new Promise(function (resolve) {
/******/ 	          var retryAttempt = 1 - retries + 1;
/******/ 	          setTimeout(function () {
/******/ 	            var retryAttemptString = '&retry-attempt=' + retryAttempt;
/******/ 	            var cacheBust = 'cache-bust=true' + retryAttemptString;
/******/ 	            queryMap[chunkId] = cacheBust;
/******/ 	            countMap[chunkId] = retries - 1;
/******/ 	            resolve(__webpack_require__.e(chunkId));
/******/ 	          }, getRetryDelay(retryAttempt));
/******/ 	        });
/******/ 	      });
/******/ 	    };
/******/ 	  }
/******/ 	})();
/******/ 	
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/index.ts ***!
  \\\\********************************************/

async function run() {
    console.log(await Promise.resolve().then(() => __webpack_require__(/*! ./async */ \\"./test/integration/fixtures/async.ts\\")));
}
run();

})();

/******/ })()
;"
`;

exports[`given config {"chunks":["main"]}, match snapshot 1`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	\\"use strict\\";
/******/ 	var __webpack_modules__ = ({

/***/ \\"./test/integration/fixtures/async.ts\\":
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/async.ts ***!
  \\\\********************************************/
/***/ ((__unused_webpack_module, exports) => {


Object.defineProperty(exports, \\"__esModule\\", ({ value: true }));
exports.foo = void 0;
exports.foo = 'foo';
console.log('foo loaded!');


/***/ })

/******/ 	});
/************************************************************************/
/******/ 	// The module cache
/******/ 	var __webpack_module_cache__ = {};
/******/ 	
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/ 		// Check if module is in cache
/******/ 		var cachedModule = __webpack_module_cache__[moduleId];
/******/ 		if (cachedModule !== undefined) {
/******/ 			return cachedModule.exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = __webpack_module_cache__[moduleId] = {
/******/ 			// no module.id needed
/******/ 			// no module.loaded needed
/******/ 			exports: {}
/******/ 		};
/******/ 	
/******/ 		// Execute the module function
/******/ 		__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/ 	
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/compat */
/******/ 	(() => {
/******/ 	  if (typeof __webpack_require__ !== 'undefined') {
/******/ 	    var oldGetScript = __webpack_require__.u;
/******/ 	    var oldLoadScript = __webpack_require__.e;
/******/ 	    var queryMap = {};
/******/ 	    var countMap = {};
/******/ 	    var getRetryDelay = function () {
/******/ 	      return 0;
/******/ 	    };
/******/ 	    __webpack_require__.u = function (chunkId) {
/******/ 	      var result = oldGetScript(chunkId);
/******/ 	      return (
/******/ 	        result +
/******/ 	        (queryMap.hasOwnProperty(chunkId) ? '?' + queryMap[chunkId] : '')
/******/ 	      );
/******/ 	    };
/******/ 	    __webpack_require__.e = function (chunkId) {
/******/ 	      var result = oldLoadScript(chunkId);
/******/ 	      return result.catch(function (error) {
/******/ 	        var retries = countMap.hasOwnProperty(chunkId) ? countMap[chunkId] : 1;
/******/ 	        if (retries < 1) {
/******/ 	          var realSrc = oldGetScript(chunkId);
/******/ 	          error.message =
/******/ 	            'Loading chunk ' +
/******/ 	            chunkId +
/******/ 	            ' failed after 1 retries.\\\\n(' +
/******/ 	            realSrc +
/******/ 	            ')';
/******/ 	          error.request = realSrc;
/******/ 	          throw error;
/******/ 	        }
/******/ 	        return new Promise(function (resolve) {
/******/ 	          var retryAttempt = 1 - retries + 1;
/******/ 	          setTimeout(function () {
/******/ 	            var retryAttemptString = '&retry-attempt=' + retryAttempt;
/******/ 	            var cacheBust = 'cache-bust=true' + retryAttemptString;
/******/ 	            queryMap[chunkId] = cacheBust;
/******/ 	            countMap[chunkId] = retries - 1;
/******/ 	            resolve(__webpack_require__.e(chunkId));
/******/ 	          }, getRetryDelay(retryAttempt));
/******/ 	        });
/******/ 	      });
/******/ 	    };
/******/ 	  }
/******/ 	})();
/******/ 	
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/index.ts ***!
  \\\\********************************************/

async function run() {
    console.log(await Promise.resolve().then(() => __webpack_require__(/*! ./async */ \\"./test/integration/fixtures/async.ts\\")));
}
run();

})();

/******/ })()
;"
`;

exports[`given config {"chunks":[]}, match snapshot 1`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	\\"use strict\\";
/******/ 	var __webpack_modules__ = ({

/***/ \\"./test/integration/fixtures/async.ts\\":
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/async.ts ***!
  \\\\********************************************/
/***/ ((__unused_webpack_module, exports) => {


Object.defineProperty(exports, \\"__esModule\\", ({ value: true }));
exports.foo = void 0;
exports.foo = 'foo';
console.log('foo loaded!');


/***/ })

/******/ 	});
/************************************************************************/
/******/ 	// The module cache
/******/ 	var __webpack_module_cache__ = {};
/******/ 	
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/ 		// Check if module is in cache
/******/ 		var cachedModule = __webpack_module_cache__[moduleId];
/******/ 		if (cachedModule !== undefined) {
/******/ 			return cachedModule.exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = __webpack_module_cache__[moduleId] = {
/******/ 			// no module.id needed
/******/ 			// no module.loaded needed
/******/ 			exports: {}
/******/ 		};
/******/ 	
/******/ 		// Execute the module function
/******/ 		__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/ 	
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/ 	
/************************************************************************/
/******/ 	/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/index.ts ***!
  \\\\********************************************/

async function run() {
    console.log(await Promise.resolve().then(() => __webpack_require__(/*! ./async */ \\"./test/integration/fixtures/async.ts\\")));
}
run();

})();

/******/ })()
;"
`;

exports[`given config undefined, match snapshot 1`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	\\"use strict\\";
/******/ 	var __webpack_modules__ = ({

/***/ \\"./test/integration/fixtures/async.ts\\":
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/async.ts ***!
  \\\\********************************************/
/***/ ((__unused_webpack_module, exports) => {


Object.defineProperty(exports, \\"__esModule\\", ({ value: true }));
exports.foo = void 0;
exports.foo = 'foo';
console.log('foo loaded!');


/***/ })

/******/ 	});
/************************************************************************/
/******/ 	// The module cache
/******/ 	var __webpack_module_cache__ = {};
/******/ 	
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/ 		// Check if module is in cache
/******/ 		var cachedModule = __webpack_module_cache__[moduleId];
/******/ 		if (cachedModule !== undefined) {
/******/ 			return cachedModule.exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = __webpack_module_cache__[moduleId] = {
/******/ 			// no module.id needed
/******/ 			// no module.loaded needed
/******/ 			exports: {}
/******/ 		};
/******/ 	
/******/ 		// Execute the module function
/******/ 		__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/ 	
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/compat */
/******/ 	(() => {
/******/ 	  if (typeof __webpack_require__ !== 'undefined') {
/******/ 	    var oldGetScript = __webpack_require__.u;
/******/ 	    var oldLoadScript = __webpack_require__.e;
/******/ 	    var queryMap = {};
/******/ 	    var countMap = {};
/******/ 	    var getRetryDelay = function () {
/******/ 	      return 0;
/******/ 	    };
/******/ 	    __webpack_require__.u = function (chunkId) {
/******/ 	      var result = oldGetScript(chunkId);
/******/ 	      return (
/******/ 	        result +
/******/ 	        (queryMap.hasOwnProperty(chunkId) ? '?' + queryMap[chunkId] : '')
/******/ 	      );
/******/ 	    };
/******/ 	    __webpack_require__.e = function (chunkId) {
/******/ 	      var result = oldLoadScript(chunkId);
/******/ 	      return result.catch(function (error) {
/******/ 	        var retries = countMap.hasOwnProperty(chunkId) ? countMap[chunkId] : 1;
/******/ 	        if (retries < 1) {
/******/ 	          var realSrc = oldGetScript(chunkId);
/******/ 	          error.message =
/******/ 	            'Loading chunk ' +
/******/ 	            chunkId +
/******/ 	            ' failed after 1 retries.\\\\n(' +
/******/ 	            realSrc +
/******/ 	            ')';
/******/ 	          error.request = realSrc;
/******/ 	          throw error;
/******/ 	        }
/******/ 	        return new Promise(function (resolve) {
/******/ 	          var retryAttempt = 1 - retries + 1;
/******/ 	          setTimeout(function () {
/******/ 	            var retryAttemptString = '&retry-attempt=' + retryAttempt;
/******/ 	            var cacheBust = 'cache-bust=true' + retryAttemptString;
/******/ 	            queryMap[chunkId] = cacheBust;
/******/ 	            countMap[chunkId] = retries - 1;
/******/ 	            resolve(__webpack_require__.e(chunkId));
/******/ 	          }, getRetryDelay(retryAttempt));
/******/ 	        });
/******/ 	      });
/******/ 	    };
/******/ 	  }
/******/ 	})();
/******/ 	
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
/*!********************************************!*\\\\
  !*** ./test/integration/fixtures/index.ts ***!
  \\\\********************************************/

async function run() {
    console.log(await Promise.resolve().then(() => __webpack_require__(/*! ./async */ \\"./test/integration/fixtures/async.ts\\")));
}
run();

})();

/******/ })()
;"
`;
